h/services/group.py

```diff
@@ -38,39 +25,52 @@ def __init__(self, session, user_fetcher, publish):
         self.user_fetcher = user_fetcher
         self.publish = publish
 
-    def create(self, name, userid, description=None, type_='private'):
+    def create_private_group(self, name, userid, description=None):
         """
-        Create a new group.
+        Create a new private group.
+
+        A private group is one that only members can read or write.
 
         :param name: the human-readable name of the group
         :param userid: the userid of the group creator
         :param description: the description of the group
-        :param type_: the type of group (private or open) which sets the
-                      appropriate access flags
 
         :returns: the created group
         """
-        creator = self.user_fetcher(userid)
-        group = Group(name=name,
-                      authority=creator.authority,
-                      creator=creator,
-                      description=description)
+        group = self._create(name=name,
+                             userid=userid,
+                             description=description,
+                             joinable_by=JoinableBy.authority,
+                             readable_by=ReadableBy.members,
+                             writeable_by=WriteableBy.members,
+                             )
+        group.members.append(group.creator)
+
+        # Flush the DB to generator group.pubid before publish()ing it.
+        self.session.flush()
 
-        access_flags = GROUP_ACCESS_FLAGS.get(type_)
-        if access_flags is None:
-            raise ValueError('Invalid group type %s' % type_)
-        for attr, value in access_flags.iteritems():
-            setattr(group, attr, value)
+        self.publish('group-join', group.pubid, group.creator.userid)
+        return group
 
-        if group.joinable_by is not None:
-            group.members.append(creator)
+    def create_open_group(self, name, userid, description=None):
+        """
+        Create a new open group.
 
-        self.session.add(group)
-        self.session.flush()
+        An open group is one that anyone can read or write.
```

Comment:
Pedantically, only those within the authority can write.

Reply:
Fixed